### PR TITLE
feat: add feature length field, annotated by add-labels function

### PR DESCRIPTION
**Pull request by nayib-jose-gloria**
#516 

Changes:
- annotate feature_length field in add-labels function, based on feature_id and pre-computed gene info CSV files (derived + committed to repo during ontology bumps). Set to 0 for non-gene features (even though a length is calculated + available for spike-in controls) 
- refactor gene ontology tool that fetches info from gene CSV files to accommodate fetching calculated feature_length 
- update tests

### CODE REVIEW

**Comment by danieljhegeman** [comment]
>for spike-in proteins

I can't believe I'm asking this but what is a "spike-in _protein_"? From Google:

```Spike-in controls are synthetic nucleic-acid sequences that are added to a user's sample and constitute internal standards for subsequent steps in the next generation sequencing workflow```

File: cellxgene_schema_cli/cellxgene_schema/ontology.py [hunk 1]
```diff
@@ -105,7 +106,22 @@ def get_symbol(self, gene_id) -> str:
         if not self.is_valid_id(gene_id):
             raise ValueError(f"The id '{gene_id}' is not a valid ENSEMBL id for '{self.species}'")
 
-        return self.gene_dict[gene_id]
+        return self.gene_dict[gene_id][0]
```
Comment (danieljhegeman) [inline cellxgene_schema_cli/cellxgene_schema/ontology.py:109]: can you elaborate on this change? I'm not seeing compensatory changes where it is called, so I'm assuming this is an unrelated bug fix? Does it need a test?

Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/ontology.py:109]: Its related, and tested in `test_ontology` as well as `test_validate`! We previously only mapped gene_id to a single value in the gene_dict (gene_label). Now that we also want to be able to fetch the gene length by ID, I updated the gene_dict to map to a tuple of (gene_label, gene_length), which we derive from the csv fixtures.

accompanying change to populating the gene_dict is here: https://github.com/chanzuckerberg/single-cell-curation/pull/624/files/fb60f0baba71b9d22da47dd6d8708ba16d75322f#diff-e03634d528c86621165785d45334e8b7255794f658cb01e118edd1f6ce817df9R71

**Comment by nayib-jose-gloria** [comment]
> > for spike-in proteins
> 
> I can't believe I'm asking this but what is a "spike-in _protein_"? From Google:
> 
> `Spike-in controls are synthetic nucleic-acid sequences that are added to a user's sample and constitute internal standards for subsequent steps in the next generation sequencing workflow`

oops--I did mean spike-in control, I have no idea why I wrote protein 😅 sorry!

File: cellxgene_schema_cli/cellxgene_schema/ontology.py (hunk 1 above)
Comment (danieljhegeman) [reply cellxgene_schema_cli/cellxgene_schema/ontology.py:109]: ah makes sense. Thanks

**Comment by danieljhegeman** [comment]
@nayib-jose-gloria tangentially-related: would you mind cleaning up this method comment, i.e. turn it into parseable English 🙏 

https://github.com/chanzuckerberg/single-cell-curation/blob/f6bb2324a42332fa1437b3d94b7f283d61837f59/cellxgene_schema_cli/cellxgene_schema/write_labels.py#L324-L325

File: cellxgene_schema_cli/cellxgene_schema/ontology.py [hunk 2]
```diff
@@ -105,7 +106,22 @@ def get_symbol(self, gene_id) -> str:
...
+
+    def get_length(self, gene_id) -> int:
```
Comment (Bento007) [inline cellxgene_schema_cli/cellxgene_schema/ontology.py:111]: type gene_id

File: cellxgene_schema_cli/cellxgene_schema/write_labels.py [hunk 3]
```diff
@@ -209,6 +209,27 @@ def _get_mapping_dict_feature_biotype(self, ids: List[str]) -> Dict[str, str]:
...
+
+        :param list[str] ids: feature IDs use for mapping
+
+        :return a mapping dictionary: {id: <int>, id: 0, ...}
+        :rtype dict
+        """
+        mapping_dict = {}
+
+        for i in ids:
+            if i.startswith("ENS"):
```
Comment (Bento007) [inline cellxgene_schema_cli/cellxgene_schema/write_labels.py:225]: Should we start putting hardcoded string into a constant.py file? Not sure we have enough cases for it to make sense.

File: cellxgene_schema_cli/tests/test_schema_compliance.py [hunk 4]
```diff
@@ -1010,6 +1010,29 @@ def test_should_warn_for_low_gene_count(self):
...
         )
 
+    def test_add_label_fields_are_reserved(self):
+        """
+        Raise an error if column names flagged as 'add_label' -> 'to_column' in the schema definition are not available.
+        """
+        for df in ["var", "raw.var"]:
+            for i in self.validator.schema_def["components"][df]["index"]["add_labels"]:
+                column = i["to_column"]
+                with self.subTest(column=column, df=df):
```
Comment (Bento007) [inline cellxgene_schema_cli/tests/test_schema_compliance.py:1020]: use https://docs.pytest.org/en/7.3.x/how-to/parametrize.html

File: cellxgene_schema_cli/tests/test_schema_compliance.py [hunk 5]
```diff
@@ -1010,6 +1010,29 @@ def test_should_warn_for_low_gene_count(self):
...
+                    # Resetting validator
+                    self.validator.adata = examples.adata.copy()
```
Comment (Bento007) [inline cellxgene_schema_cli/tests/test_schema_compliance.py:1022]: I believe these 2 lines can be removed if you switch to a pytest approach.

**Review by Bento007 — CHANGES_REQUESTED** [review]

File: cellxgene_schema_cli/cellxgene_schema/write_labels.py (hunk 3 above)
Comment (danieljhegeman) [reply cellxgene_schema_cli/cellxgene_schema/write_labels.py:225]: Might be a good thing to do at the end of the epic? Or we can start now

File: cellxgene_schema_cli/cellxgene_schema/ontology.py [hunk 6]
```diff
@@ -105,7 +106,22 @@ def get_symbol(self, gene_id) -> str:
...
+        """
+        Gets feature length associated to the ENSEBML id
+
+        :param str gene_id: ENSEMBL gene id
+
+        :rtype int
+        :return A gene length
+        """
+
+        if not self.is_valid_id(gene_id):
```
Comment (danieljhegeman) [inline cellxgene_schema_cli/cellxgene_schema/ontology.py:121]: @Bento007 has encouraged me to avoid negatives. 

Could turn this to a one-liner:
```return self.gene_dict[gene_id][1] if self.is_valid_id(gene_id) else raise ValueError(f"The id '{gene_id}' is not a valid ENSEMBL id for '{self.species}'")
```

🙂

File: cellxgene_schema_cli/tests/test_validate.py [hunk 7]
```diff
@@ -136,7 +135,29 @@ def test_get_dictionary_mapping_feature_reference(self):
...
+            0,  # non-gene feature, so set to 0 regardless of csv value
+            42738,
+            4045,
+            3822,
+        ]
+        expected_dict = dict(zip(ids, gene_lengths))
+        self.assertEqual(self.writer._get_mapping_dict_feature_length(ids), expected_dict)
+
+        # Bad
+        ids = ["NO_GENE"]
```
Comment (danieljhegeman) [inline cellxgene_schema_cli/tests/test_validate.py:160]: For all of the `test_get_dictionary_*` tests in `TestAddLabelFunctions`, you wish to consider factoring out common dict comparison logic and `ids` initial variables. Could create a helper function or could simply group all under a single test. Not required though.

**Review by danieljhegeman — APPROVED** [review]

File: cellxgene_schema_cli/cellxgene_schema/write_labels.py (hunk 3 above)
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/write_labels.py:225]: I don't think it's necessary at this point, we can revisit if we find that it's coming up frequently and causing confusion.

File: cellxgene_schema_cli/tests/test_schema_compliance.py (hunk 4 above)
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/tests/test_schema_compliance.py:1020]: https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/gh/chanzuckerberg/single-cell-curation/633 tracking a separate issue for this

**Review by Bento007 — APPROVED** [review]
